cores/arduino/SoftwareSerial.cpp

```diff
@@ -0,0 +1,420 @@
+uint32_t SoftwareSerial::cur_speed = 0;
+
+//
+// Private methods
+//
+
+void SoftwareSerial::setSpeed(uint32_t speed)
+{
+  if (speed != cur_speed) {
+    timer.pause();
+    if (speed != 0) {
+      // Disable the timer
+      uint32_t clock_rate, cmp_value;
+      // Get timer clock
+      clock_rate = timer.getTimerClkFreq();
+      int pre = 1;
+      // Calculate prescale an compare value
+      do {
+        cmp_value = clock_rate / (speed * OVERSAMPLE);
+        if (cmp_value >= UINT16_MAX) {
+          clock_rate = clock_rate / 2;
+          pre *= 2;
+        }
+      } while (cmp_value >= UINT16_MAX);
+      timer.setPrescaleFactor(pre);
+      timer.setOverflow(cmp_value);
+      timer.setCount(0);
+      timer.attachInterrupt(&handleInterrupt);
+      timer.resume();
+    } else {
+      timer.detachInterrupt();
+    }
+    cur_speed = speed;
+  }
+}
+
+// This function sets the current object as the "listening"
+// one and returns true if it replaces another
+bool SoftwareSerial::listen()
+{
+  if (_receivePin >= 0) {
+    // wait for any transmit to complete as we may change speed
+    while (active_out)
+      ;
+    if (active_listener) {
+      active_listener->stopListening();
+    }
+    rx_tick_cnt = 1;
+    rx_bit_cnt = -1;
+    setSpeed(_speed);
+    active_listener = this;
+    if (!_half_duplex) {
+      active_in = this;
+    }
+    return true;
+  }
+  return false;
+}
+
+// Stop listening. Returns true if we were actually listening.
+bool SoftwareSerial::stopListening()
+{
+  if (active_listener == this) {
+    // wait for any output to complete
+    while (active_out)
+      ;
+    if (_half_duplex) {
+      setRXTX(false);
+    }
+    active_listener = nullptr;
+    active_in = nullptr;
+    // turn off ints
+    setSpeed(0);
+    return true;
+  }
+  return false;
+}
+
+inline void SoftwareSerial::setTX()
+{
+  digitalWriteFast(digitalPinToPinName(_transmitPin), _inverse_logic ? LOW : HIGH);
+  pinMode(_transmitPin, OUTPUT);
+}
+
+inline void SoftwareSerial::setRX()
+{
+  if (_receivePin > 0) {
+    pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+  }
+}
+
+inline void SoftwareSerial::setRXTX(bool input)
+{
+  if (_half_duplex) {
+    if (input) {
+      if (active_in != this) {
+        setRX();
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 2;
+        active_in = this;
+      }
+    } else {
+      if (active_in == this) {
+        setTX();
+        active_in = nullptr;
+      }
+    }
+  }
+}
+
+inline void SoftwareSerial::send()
+{
+  if (--tx_tick_cnt <= 0) {
+    if (tx_bit_cnt++ < 10) {
+      // send data (including start and stop bits)
+      digitalWriteFast(digitalPinToPinName(_transmitPin), (tx_buffer & 1) ? ~_inverse_logic : _inverse_logic);
+      tx_buffer >>= 1;
+      tx_tick_cnt = OVERSAMPLE;
+    } else {
+      tx_tick_cnt = 1;
+      if (_output_pending) {
+        active_out = nullptr;
+      } else if (tx_bit_cnt > 10 + OVERSAMPLE * 5) {
```

Comment:
This is coming from the original implementation for lpc176x. It seems a pause is required when in half-duplex mode before switching from TX to RX mode. However, the pause here is OVERSAMPLE * 5 bits, which cannot be right. I could be that 5 bits is desired here, but I must admit I have not tested the half-duplex mode. I can remove the half-duplex code completely, or change this to 5 bits delay (and change the condition so the delay is really only there in half-duplex mode)

Reply:
If 5 bit is desired then OVERSAMPLE * 5 bits is good (OVERSAMPLE correspond to 1 bits)

Reply:
I don't think so. We're comparing tx_bit_cnt here, not tx_tick_cnt.

Reply:
Yes, but after the 10 bit are send, tx_tick_cnt is no more set to OVERSAMPLE. And so , once tx_tick_cnt reach negative value, it will remain negative. *
So  'if (--tx_tick_cnt <= 0)' will match at each interrupt. Then tx_bit_cnt will be incremented at each interrupt.

Reply:
Right, I didn't catch that one. I will change it.